File: prototype/rails/src/test/domain/reports/activity_timeline_test.rb
require "test_helper"

module Domain
  module Reports
    class ActivityTimelineTest < ActiveSupport::TestCase
      test "it returns one row per day oldest first" do
        days = ActivityTimeline.last_days({}, ends_on: Time.new(2026, 8, 22, 17, 30, 0), days: 3)

        assert_equal 3, days.length
        assert_equal [Date.new(2026, 8, 20), Date.new(2026, 8, 21), Date.new(2026, 8, 22)], days.map(&:date)
      end

      test "it reads the counts recorded for a day" do
        counts = { Date.new(2026, 8, 21) => { "view" => 4, "favorite" => 1, "cart_add" => 2 } }

        days = ActivityTimeline.last_days(counts, ends_on: Date.new(2026, 8, 22), days: 2)

        assert_equal 4, days[0].totals.views
        assert_equal 1, days[0].totals.favorites
        assert_equal 2, days[0].totals.cart_adds
      end

      test "a day with no events counts zero" do
        counts = { Date.new(2026, 8, 22) => { "view" => 7 } }

        days = ActivityTimeline.last_days(counts, ends_on: Date.new(2026, 8, 22), days: 2)

        assert_equal 0, days[0].totals.total
        assert_equal 7, days[1].totals.views
      end

      test "it ignores counts outside the window" do
        counts = { Date.new(2026, 7, 1) => { "view" => 99 } }

        days = ActivityTimeline.last_days(counts, ends_on: Date.new(2026, 8, 22), days: 2)

        assert_equal 0, days.sum { |day| day.totals.total }
      end

      test "a fortnight ends on the day asked for" do
        days = ActivityTimeline.last_days({}, ends_on: Date.new(2026, 8, 22), days: 14)

        assert_equal Date.new(2026, 8, 9), days.first.date
        assert_equal Date.new(2026, 8, 22), days.last.date
      end

      test "it refuses a window shorter than a day" do
        assert_raises(ArgumentError) { ActivityTimeline.last_days({}, ends_on: Date.new(2026, 8, 22), days: 0) }
      end
    end
  end
end

File: prototype/node/src/app/core/reports/activity-timeline.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { activityTimeline } from './activity-timeline.ts'

test('it returns one row per day, oldest first', () => {
  const days = activityTimeline({}, { endsOn: new Date('2026-08-22T17:30:00.000Z'), days: 3 })

  assert.deepEqual(
    days.map((day) => day.day),
    ['2026-08-20', '2026-08-21', '2026-08-22'],
  )
})

test('it reads the counts recorded for a day', () => {
  const counts = { '2026-08-21': { view: 4, favorite: 1, cart_add: 2 } }

  const days = activityTimeline(counts, { endsOn: new Date('2026-08-22T00:00:00.000Z'), days: 2 })

  assert.equal(days[0]?.totals.views, 4)
  assert.equal(days[0]?.totals.favorites, 1)
  assert.equal(days[0]?.totals.cartAdds, 2)
})

test('a day with no events counts zero', () => {
  const counts = { '2026-08-22': { view: 7 } }

  const days = activityTimeline(counts, { endsOn: new Date('2026-08-22T00:00:00.000Z'), days: 2 })

  assert.equal(days[0]?.totals.views, 0)
  assert.equal(days[1]?.totals.views, 7)
})

test('it ignores counts outside the window', () => {
  const counts = { '2026-07-01': { view: 99 } }

  const days = activityTimeline(counts, { endsOn: new Date('2026-08-22T00:00:00.000Z'), days: 2 })

  assert.equal(days.reduce((total, day) => total + day.totals.views, 0), 0)
})

test('a fortnight ends on the day asked for', () => {
  const days = activityTimeline({}, { endsOn: new Date('2026-08-22T00:00:00.000Z'), days: 14 })

  assert.equal(days[0]?.day, '2026-08-09')
  assert.equal(days.at(-1)?.day, '2026-08-22')
})

test('it refuses a window shorter than a day', () => {
  assert.throws(
    () => activityTimeline({}, { endsOn: new Date('2026-08-22T00:00:00.000Z'), days: 0 }),
    RangeError,
  )
})
